ci-operator/step-registry/github/branch-sync/github-branch-sync-commands.sh

```diff
@@ -0,0 +1,129 @@
+echo "🔐 Generating JWT…"
+# create the JWT needed to get an app install token needed for API requests
+# the token is short-lived and will expire after EXP below
+NOW=$(date +%s)
+EXP=$((NOW + 600))
+HEADER=$(printf '{"alg":"RS256","typ":"JWT"}' | b64url)
+PAYLOAD=$(printf '{"iat":%d,"exp":%d,"iss":%d}' "$NOW" "$EXP" "$GITHUB_APP_ID" | b64url)
+SIG_INPUT="$HEADER.$PAYLOAD"
+
+echo "🖋 Signing JWT…"
+SIGNATURE=$(
+  printf '%s' "$SIG_INPUT" \
+    | openssl dgst -sha256 -sign "$GITHUB_APP_PRIVATE_KEY_FILE" \
+    | b64url
+)
+JWT="$HEADER.$PAYLOAD.$SIGNATURE"
+
+echo "🔗 Exchanging JWT for installation token…"
+INSTALLATION_ID=$(curl -sS -H "Authorization: Bearer $JWT" -H "Accept: application/vnd.github+json" \
+    "https://api.github.com/repos/${REPO}/installation" | jq -r .id)
+GITHUB_TOKEN=$(curl -sS -H "Authorization: Bearer $JWT" -H "Accept: application/vnd.github+json" -X POST \
+    "https://api.github.com/app/installations/${INSTALLATION_ID}/access_tokens" | jq -r .token)
+echo "✅ Received install token (ID: ${INSTALLATION_ID})"
+
+echo "📥 Cloning repo and setting up remotes…"
+# get the repo
+WORKDIR="$(mktemp -d)"
+cd "$WORKDIR"
+git clone --single-branch --branch "${SOURCE_BRANCH}" "https://github.com/${REPO}" repo
+cd repo
+git remote set-url origin "https://x-access-token:${GITHUB_TOKEN}@github.com/${REPO}"
+git fetch origin \
+  "+refs/heads/${SOURCE_BRANCH}:refs/remotes/origin/${SOURCE_BRANCH}" \
+  "+refs/heads/${TARGET_BRANCH}:refs/remotes/origin/${TARGET_BRANCH}"
+
+TITLE_DATE=$(date +%m-%d-%Y)
+BRANCH="sync-${SOURCE_BRANCH}-to-${TARGET_BRANCH}-${TITLE_DATE}"
+PR_TITLE="NO-JIRA: Branch Sync ${SOURCE_BRANCH} to ${TARGET_BRANCH} [${TITLE_DATE}]"
+PR_BODY="Automated branch sync: ${SOURCE_BRANCH} to ${TARGET_BRANCH}."
+
+# exit if no new commits are available to sync
+NEW_COMMITS=$(git rev-list "origin/${TARGET_BRANCH}..origin/${SOURCE_BRANCH}" --count)
+(( NEW_COMMITS == 0 )) && { echo "No changes to sync; exiting."; exit 0; }
+
+# exit if there is already an open PR for branch sync
+OPEN=$(curl -sS -H "Authorization: token ${GITHUB_TOKEN}" \
+  "https://api.github.com/repos/${REPO}/pulls?state=open&base=${TARGET_BRANCH}&per_page=100" \
+  | jq -r --arg tb "$TARGET_BRANCH" --arg sb "$SOURCE_BRANCH" '
+  .[] | select(.title | test("\\[" + $tb + "\\] DownStream Merge Sync from " + $sb + " \\[[0-9]{2}-[0-9]{2}-[0-9]{4}\\]")) | .number' \
+  | head -n1
+)
+
+if [[ -n "$OPEN" ]]; then
+  echo "Open branch-sync PR #$OPEN; exiting."
+  exit 0
+fi
+
+echo "🧹 Deleting stale branch ${BRANCH}, if any…"
+# if an earlier failed run left the branch behind, delete it now
+curl -sS -H "Authorization: token ${GITHUB_TOKEN}" -X DELETE \
+  "https://api.github.com/repos/${REPO}/git/refs/heads/${BRANCH}" || true
+
+# create branch off TARGET, merge SOURCE with -X theirs
+git checkout -b "$BRANCH" "origin/${TARGET_BRANCH}"
+CONFLICT=false
+if ! git merge -X theirs "origin/${SOURCE_BRANCH}" --no-edit; then
+  echo "Merge conflict (even with -X theirs)."
```

Comment:
I wonder what happens if:

1. Someone merges a manual backport due to urgency facing conflicts resolved manually
2. Then this automated backport is done

Are we safe using `X -theirs` in this situation? I am not sure in which cases `X -theirs` would report conflicts and I am not sure if we would be safe.
If we aren't safe, I think we need to mandate all manual backports to require no conflicts, which means mandating pulling in all the commits that are needed to avoid them. I wouldn't assume this is mandate is in place. I think it would be good to have it even if it causes no trouble for the sync process.

Reply:
what I have learned is that -X theirs can still have conflicts with some other cases like filename changes, etc. I haven't seen it yet, but have this hear to be safe.

my thinking is that we do want 4.x-1 and 4.x-2 to be identical to 4.x so -X theirs should guarantee that.

if there was a manual back-port with trouble then we hopefully catch that when it's created and reviewed and we will want whatever we did to make it right to end up in the branch syncs anyway. That can be from your idea (mandating pulling all the commits) or whatever else we decide is required (e.g., manually fixing the conflicts one by one)